source_catalog unit tests fail with dev photutils

_Issue [JP-2203](https://jira.stsci.edu/browse/JP-2203) was created on JIRA by [James Davies](https://jira.stsci.edu/secure/ViewProfile.jspa?name=jdavies):_

In `jwst` CI on the master branch with dev dependencies, including `photutils` we are seeing 2 `source_catalog` unit tests fail.

<https://github.com/spacetelescope/jwst/runs/3079494003?check_suite_focus=true#step:8:1483>

```
$ pip install -r requirements-dev.txt
$ pytest jwst/source_catalog
============================== test session starts ==============================
platform darwin -- Python 3.9.5, pytest-6.2.4, py-1.10.0, pluggy-0.13.1
crds_context: jwst_0738.pmap
rootdir: /Users/jdavies/dev/jwst, configfile: setup.cfg
plugins: cov-2.12.1, doctestplus-0.10.0, requests-mock-1.9.3, asdf-2.9.0.dev6+g887c5be, ci-watson-0.5, openfiles-0.5.0, jwst-1.2.4.dev20+g8a8f148d
collected 3 items                                                               

jwst/source_catalog/tests/test_source_catalog.py FF.                      [100%]

=================================== FAILURES ====================================
___________________________ test_source_catalog[5-2] ____________________________

self = <jwst.source_catalog.source_catalog.JWSTSourceCatalog object at 0x7ff888da2b20>

    def set_segment_properties(self):
        """
        Calculate the segment-based source photometry and morphologies.
    
        The values are set as dynamic attributes.
        """
        segm_cat = SourceCatalog(self.model.data, self.segment_img,
                                 error=self.model.err, kernel=self.kernel,
                                 wcs=self.wcs)
    
        self._xpeak = segm_cat.maxval_xindex
        self._ypeak = segm_cat.maxval_yindex
    
        # rename some columns in the output catalog
        prop_names = {}
        prop_names['isophotal_flux'] = 'segment_flux'
        prop_names['isophotal_flux_err'] = 'segment_fluxerr'
        prop_names['isophotal_area'] = 'area'
    
        for column in self.segment_colnames:
            # define the property name
            prop_name = prop_names.get(column, column)
            try:
>               value = getattr(segm_cat, prop_name)
E               AttributeError: 'SourceCatalog' object has no attribute 'sky_orientation'

jwst/source_catalog/source_catalog.py:631: AttributeError

During handling of the above exception, another exception occurred:

nircam_model = <ImageModel(101, 101)>, npixels = 5, nsources = 2

    @pytest.mark.parametrize('npixels, nsources', ((5, 2), (1000, 1), (5000, 0)))
    def test_source_catalog(nircam_model, npixels, nsources):
    
        step = SourceCatalogStep(snr_threshold=0.5, npixels=npixels,
                                 bkg_boxsize=50, kernel_fwhm=2.0,
                                 save_results=False)
>       cat = step.run(nircam_model)

jwst/source_catalog/tests/test_source_catalog.py:58: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
../../miniconda3/envs/test/lib/python3.9/site-packages/stpipe/step.py:407: in run
    step_result = self.process(*args)
jwst/source_catalog/source_catalog_step.py:109: in process
    catalog = catobj.catalog
../../miniconda3/envs/test/lib/python3.9/site-packages/astropy/utils/decorators.py:767: in __get__
    val = self.fget(obj)
jwst/source_catalog/source_catalog.py:1300: in catalog
    self.set_segment_properties()
jwst/source_catalog/source_catalog.py:633: in set_segment_properties
    value = getattr(self, prop_name)
../../miniconda3/envs/test/lib/python3.9/site-packages/astropy/utils/decorators.py:767: in __get__
    val = self.fget(obj)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <jwst.source_catalog.source_catalog.JWSTSourceCatalog object at 0x7ff888da2b20>

    @lazyproperty
    def sky_orientation(self):
        """
        The orientation of the source major axis as the position angle
        in degrees measured East of North.
        """
        # NOTE: crpix1 and crpix2 are 1-based values
        skycoord = self.wcs.pixel_to_world(self.model.meta.wcsinfo.crpix1 - 1,
                                           self.model.meta.wcsinfo.crpix2 - 1)
>       _, angle = _pixel_scale_angle_at_skycoord(skycoord, self.wcs)
E       ValueError: too many values to unpack (expected 2)

jwst/source_catalog/source_catalog.py:691: ValueError
__________________________ test_source_catalog[1000-1] __________________________

self = <jwst.source_catalog.source_catalog.JWSTSourceCatalog object at 0x7ff89891c9d0>

    def set_segment_properties(self):
        """
        Calculate the segment-based source photometry and morphologies.
    
        The values are set as dynamic attributes.
        """
        segm_cat = SourceCatalog(self.model.data, self.segment_img,
                                 error=self.model.err, kernel=self.kernel,
                                 wcs=self.wcs)
    
        self._xpeak = segm_cat.maxval_xindex
        self._ypeak = segm_cat.maxval_yindex
    
        # rename some columns in the output catalog
        prop_names = {}
        prop_names['isophotal_flux'] = 'segment_flux'
        prop_names['isophotal_flux_err'] = 'segment_fluxerr'
        prop_names['isophotal_area'] = 'area'
    
        for column in self.segment_colnames:
            # define the property name
            prop_name = prop_names.get(column, column)
            try:
>               value = getattr(segm_cat, prop_name)
E               AttributeError: 'SourceCatalog' object has no attribute 'sky_orientation'

jwst/source_catalog/source_catalog.py:631: AttributeError

During handling of the above exception, another exception occurred:

nircam_model = <ImageModel(101, 101)>, npixels = 1000, nsources = 1

    @pytest.mark.parametrize('npixels, nsources', ((5, 2), (1000, 1), (5000, 0)))
    def test_source_catalog(nircam_model, npixels, nsources):
    
        step = SourceCatalogStep(snr_threshold=0.5, npixels=npixels,
                                 bkg_boxsize=50, kernel_fwhm=2.0,
                                 save_results=False)
>       cat = step.run(nircam_model)

jwst/source_catalog/tests/test_source_catalog.py:58: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
../../miniconda3/envs/test/lib/python3.9/site-packages/stpipe/step.py:407: in run
    step_result = self.process(*args)
jwst/source_catalog/source_catalog_step.py:109: in process
    catalog = catobj.catalog
../../miniconda3/envs/test/lib/python3.9/site-packages/astropy/utils/decorators.py:767: in __get__
    val = self.fget(obj)
jwst/source_catalog/source_catalog.py:1300: in catalog
    self.set_segment_properties()
jwst/source_catalog/source_catalog.py:633: in set_segment_properties
    value = getattr(self, prop_name)
../../miniconda3/envs/test/lib/python3.9/site-packages/astropy/utils/decorators.py:767: in __get__
    val = self.fget(obj)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <jwst.source_catalog.source_catalog.JWSTSourceCatalog object at 0x7ff89891c9d0>

    @lazyproperty
    def sky_orientation(self):
        """
        The orientation of the source major axis as the position angle
        in degrees measured East of North.
        """
        # NOTE: crpix1 and crpix2 are 1-based values
        skycoord = self.wcs.pixel_to_world(self.model.meta.wcsinfo.crpix1 - 1,
                                           self.model.meta.wcsinfo.crpix2 - 1)
>       _, angle = _pixel_scale_angle_at_skycoord(skycoord, self.wcs)
E       ValueError: too many values to unpack (expected 2)

jwst/source_catalog/source_catalog.py:691: ValueError
============================ short test summary info ============================
FAILED jwst/source_catalog/tests/test_source_catalog.py::test_source_catalog[5-2]
FAILED jwst/source_catalog/tests/test_source_catalog.py::test_source_catalog[1000-1]
========================== 2 failed, 1 passed in 3.27s ==========================
```

So updates to either the code or test code will be needed.

--- a/CHANGES.rst
+++ b/CHANGES.rst
@@ -51,6 +51,11 @@ ramp_fitting
 - Update ``RampFitStep`` to pass DQ flags as a parameter to the ``ramp_fit``
   algorithm code in stcal.  Bump version requirement for stcal.  [#6072]
 
+source_catalog
+--------------
+
+- Fixed issue with non-finite positions for aperture photometry. [#6206]
+
 wfss_contam
 -----------
 

--- a/jwst/source_catalog/_wcs_helpers.py
+++ b/jwst/source_catalog/_wcs_helpers.py
@@ -1,0 +1,65 @@
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+# (taken from photutils: should probably migrate into astropy.wcs)
+"""
+This module provides WCS helper tools.
+"""
+
+import astropy.units as u
+import numpy as np
+
+
+def pixel_scale_angle_at_skycoord(skycoord, wcs, offset=1 * u.arcsec):
+    """
+    Calculate the pixel coordinate and scale and WCS rotation angle at
+    the position of a SkyCoord coordinate.
+
+    Parameters
+    ----------
+    skycoord : `~astropy.coordinates.SkyCoord`
+        The SkyCoord coordinate.
+
+    wcs : WCS object
+        A world coordinate system (WCS) transformation that
+        supports the `astropy shared interface for WCS
+        <https://docs.astropy.org/en/stable/wcs/wcsapi.html>`_ (e.g.,
+        `astropy.wcs.WCS`, `gwcs.wcs.WCS`).
+
+    offset : `~astropy.units.Quantity`
+        A small angular offset to use to compute the pixel scale and
+        position angle.
+
+    Returns
+    -------
+    xypos : tuple of float
+        The (x, y) pixel coordinate.
+
+    scale : `~astropy.units.Quantity`
+        The pixel scale in arcsec/pixel.
+
+    angle : `~astropy.units.Quantity`
+        The angle (in degrees) measured counterclockwise from the
+        positive x axis to the "North" axis of the celestial coordinate
+        system.
+
+    Notes
+    -----
+    If distortions are present in the image, the x and y pixel scales
+    likely differ.  This function computes a single pixel scale along
+    the North/South axis.
+    """
+    # Convert to pixel coordinates
+    xpos, ypos = wcs.world_to_pixel(skycoord)
+
+    # We take a point directly North (i.e., latitude offset) the
+    # input sky coordinate and convert it to pixel coordinates,
+    # then we use the pixel deltas between the input and offset sky
+    # coordinate to calculate the pixel scale and angle.
+    skycoord_offset = skycoord.directional_offset_by(0.0, offset)
+    x_offset, y_offset = wcs.world_to_pixel(skycoord_offset)
+
+    dx = x_offset - xpos
+    dy = y_offset - ypos
+    scale = offset.to(u.arcsec) / (np.hypot(dx, dy) * u.pixel)
+    angle = (np.arctan2(dy, dx) * u.radian).to(u.deg)
+
+    return (xpos, ypos), scale, angle

--- a/jwst/source_catalog/source_catalog.py
+++ b/jwst/source_catalog/source_catalog.py
@@ -644,6 +644,20 @@ class JWSTSourceCatalog:
         return np.transpose((self.xcentroid, self.ycentroid))
 
     @lazyproperty
+    def _xypos_finite(self):
+        """
+        The (x, y) source positions, where non-finite positions are
+        set to a large negative value.
+
+        At this position the aperture will not overlap the data, thus
+        returning NaN fluxes and errors.
+        """
+        xypos = self.xypos.copy()
+        nanmask = ~np.isfinite(xypos)
+        xypos[nanmask] = -1000.
+        return xypos
+
+    @lazyproperty
     def _isophotal_abmag(self):
         """
         The isophotal AB magnitude and error.
@@ -833,18 +847,18 @@ class JWSTSourceCatalog:
         median, sqrt(pi / 2N) * std.
         """
         bkg_aper = CircularAnnulus(
-            self.xypos, self.aperture_params['bkg_aperture_inner_radius'],
+            self._xypos_finite,
+            self.aperture_params['bkg_aperture_inner_radius'],
             self.aperture_params['bkg_aperture_outer_radius'])
         bkg_aper_masks = bkg_aper.to_mask(method='center')
-        sigclip = SigmaClip(sigma=3)
+        sigclip = SigmaClip(sigma=3.)
 
         nvalues = []
         bkg_median = []
         bkg_std = []
         for mask in bkg_aper_masks:
-            bkg_data = mask.multiply(self.model.data.value)
-            bkg_data_1d = bkg_data[mask.data > 0]
-            values = sigclip(bkg_data_1d, masked=False)
+            bkg_data = mask.get_values(self.model.data.value)
+            values = sigclip(bkg_data, masked=False)
             nvalues.append(values.size)
             bkg_median.append(np.median(values))
             bkg_std.append(np.std(values))
@@ -879,7 +893,7 @@ class JWSTSourceCatalog:
 
         The values are set as dynamic attributes.
         """
-        apertures = [CircularAperture(self.xypos, radius) for radius in
+        apertures = [CircularAperture(self._xypos_finite, radius) for radius in
                      self.aperture_params['aperture_radii']]
         aper_phot = aperture_photometry(self.model.data, apertures,
                                         error=self.model.err)

--- a/jwst/source_catalog/source_catalog.py
+++ b/jwst/source_catalog/source_catalog.py
@@ -13,6 +13,7 @@ from astropy.stats import gaussian_fwhm_to_sigma, SigmaClip
 from astropy.table import QTable
 import astropy.units as u
 from astropy.utils import lazyproperty
+from astropy.utils.exceptions import AstropyUserWarning
 import numpy as np
 from scipy import __version__ as scipy_version
 from scipy import ndimage
@@ -547,7 +548,7 @@ class JWSTSourceCatalog:
         """
         # ignore RunTimeWarning if flux or flux_err contains NaNs
         with warnings.catch_warnings():
-            warnings.simplefilter("ignore", category=RuntimeWarning)
+            warnings.simplefilter('ignore', category=RuntimeWarning)
 
             abmag = -2.5 * np.log10(flux.value) + 8.9
             abmag_err = 2.5 * np.log10(1.0 + (flux_err.value / flux.value))
@@ -853,15 +854,19 @@ class JWSTSourceCatalog:
         bkg_aper_masks = bkg_aper.to_mask(method='center')
         sigclip = SigmaClip(sigma=3.)
 
-        nvalues = []
-        bkg_median = []
-        bkg_std = []
-        for mask in bkg_aper_masks:
-            bkg_data = mask.get_values(self.model.data.value)
-            values = sigclip(bkg_data, masked=False)
-            nvalues.append(values.size)
-            bkg_median.append(np.median(values))
-            bkg_std.append(np.std(values))
+        with warnings.catch_warnings():
+            warnings.simplefilter('ignore', category=RuntimeWarning)
+            warnings.simplefilter('ignore', category=AstropyUserWarning)
+
+            nvalues = []
+            bkg_median = []
+            bkg_std = []
+            for mask in bkg_aper_masks:
+                bkg_data = mask.get_values(self.model.data.value)
+                values = sigclip(bkg_data, masked=False)
+                nvalues.append(values.size)
+                bkg_median.append(np.median(values))
+                bkg_std.append(np.std(values))
 
         nvalues = np.array(nvalues)
         bkg_median = np.array(bkg_median)

--- a/jwst/source_catalog/source_catalog.py
+++ b/jwst/source_catalog/source_catalog.py
@@ -1163,7 +1163,11 @@ class JWSTSourceCatalog:
         """
         if len(self._ckdtree_query[1]) == 1:  # only one detected source
             return np.nan
-        return self.label[self._ckdtree_query[1]]
+
+        idx = self._ckdtree_query[1].copy()
+        mask = idx >= len(self.label)
+        idx[mask] = 0
+        return np.where(mask, self.label, self.label[idx])
 
     @lazyproperty
     def nn_dist(self):

--- a/jwst/source_catalog/source_catalog.py
+++ b/jwst/source_catalog/source_catalog.py
@@ -25,12 +25,12 @@ from photutils.segmentation import (detect_sources, deblend_sources,
                                     SourceCatalog)
 from photutils.aperture import (CircularAperture, CircularAnnulus,
                                 aperture_photometry)
-from photutils.utils._wcs_helpers import _pixel_scale_angle_at_skycoord
 
 from jwst import __version__ as jwst_version
 
 from .. import datamodels
 from ..datamodels import ImageModel, ABVegaOffsetModel
+from ._wcs_helpers import pixel_scale_angle_at_skycoord
 
 log = logging.getLogger(__name__)
 log.setLevel(logging.DEBUG)
@@ -703,7 +703,7 @@ class JWSTSourceCatalog:
         # NOTE: crpix1 and crpix2 are 1-based values
         skycoord = self.wcs.pixel_to_world(self.model.meta.wcsinfo.crpix1 - 1,
                                            self.model.meta.wcsinfo.crpix2 - 1)
-        _, angle = _pixel_scale_angle_at_skycoord(skycoord, self.wcs)
+        _, _, angle = pixel_scale_angle_at_skycoord(skycoord, self.wcs)
 
         return (180.0 * u.deg) - angle + self.orientation
 

